nixos/nebula: add CAP_NET_BIND_SERVICE when lighthouse node serves DNS

siriobalmelli opened the pull request:
Serving DNS fails in the absence of CAP_NET_BIND_SERVICE.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

siriobalmelli commented:
@numinit I would appreciate your review on this 🙏

numinit commented on nixos/modules/services/networking/nebula.nix:
```diff
@@ -210,6 +210,11 @@ in
           );
+        capabilities = concatStringsSep " " ([
+          "CAP_NET_ADMIN"
+        ] ++ (optionals (settings.lighthouse.serve_dns or false) [
```
This condition should probably be:

```
let
  nebulaPort = settings.listen.port;
  dnsPort = if (settings.lighthouse.serve_dns or false) then settings.lighthouse.dns.port or -1 else -1;
in nebulaPort > 0 && nebulaPort < 1024 || dnsPort > 0 && dnsPort < 1024;
```

Do you mind adding `services.nebula.networks.<name>.dns.{enable,port,host}` as well? The port should likely default to 5353 so we don't need the extra capability.

numinit commented:
May be worth updating the NixOS test too.

numinit commented:
Hey, thanks for taking this btw! Let me know if you need any help getting it across the finish line.

siriobalmelli commented:
Thanks :)

Got it done and working well on my own nebula network.

A bit stumped about how to extend the tests to cover this however, any help there is appreciated.

numinit commented on nixos/modules/services/networking/nebula.nix:
```diff
@@ -231,6 +249,10 @@ in
               settings
           );
+          capabilities = lib.concatStringsSep " " (
+            (lib.optional (!settings.tun.disabled) "CAP_NET_ADMIN")
```
You can get rid of the parens around (lib.optional ..) - ++ is higher precedence than the function call.

siriobalmelli replied:
Thank you, TIL

numinit commented on nixos/modules/services/networking/nebula.nix:
```diff
@@ -84,6 +84,24 @@ in
+
+              lighthouse.dns.port = lib.mkOption {
+                type = lib.types.nullOr lib.types.port;
+                default = 53;
```
Maybe default this to something like `5353`

numinit commented:
For the tests, maybe make a `dig` query to the other hosts in the network that we test. It's mostly optional though.

numinit commented on nixos/modules/services/networking/nebula.nix:
```diff
@@ -84,6 +84,24 @@ in
+
+              lighthouse.dns.host = lib.mkOption {
+                type = lib.types.str;
+                default = "0.0.0.0";
```
Better default is possibly localhost.

Jaculabilis replied:
I agree `localhost` is better as a "secure defaults" choice, so the default doesn't listen on public interfaces. The description may want a callout, though, or have an example value of some nebula address like "10.0.0.5".

numinit commented:
NixOS tests all pass, looks good except for the above unless @Jaculabilis has any more changes. Thanks for the contribution!

numinit commented on nixos/tests/nebula.nix:
```diff
@@ -338,6 +349,8 @@ import ./make-test-python.nix (
         allowAny.succeed("ping -c3 10.0.100.1")
         allowAny.fail("ping -c3 10.0.100.3")
+        # allowAny can also resolve DNS on lighthouse
+        allowAny.succeed("dig @10.0.100.1 allowToLighthouse")
```
`dig` returns 0 if there's a NXDOMAIN. Recommend `dig @10.0.100.1 allowToLighthouse | grep -E 'allowToLighthouse\.\s+[0-9]+\s+IN\s+A\s+10\.0\.100\.4'`

siriobalmelli commented:
Respun with requested changes. Unable to run nixos tests at the moment to verify DNS grep, apologies.

numinit commented:
Indeed it works :-)

```
allowAny: must succeed: dig @10.0.100.1 allowToLighthouse | tee /dev/stderr | grep -E 'allowToLighthouse\.\s+[0-9]+\s+IN\s+A\s+10\.0\.100\.4'
allowAny # 
allowAny # ; <<>> DiG 9.18.28 <<>> @10.0.100.1 allowToLighthouse
allowAny # ; (1 server found)
allowAny # ;; global options: +cmd
allowAny # ;; Got answer:
allowAny # ;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 41232
allowAny # ;; flags: qr rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
allowAny # ;; WARNING: recursion requested but not available
allowAny # 
allowAny # ;; QUESTION SECTION:
allowAny # ;allowToLighthouse.          IN      A
allowAny # 
allowAny # ;; ANSWER SECTION:
allowAny # allowToLighthouse.   3600    IN      A       10.0.100.4
allowAny # 
allowAny # ;; Query time: 3 msec
allowAny # ;; SERVER: 10.0.100.1#53(10.0.100.1) (UDP)
allowAny # ;; WHEN: Mon Jan 20 07:12:50 UTC 2025
allowAny # ;; MSG SIZE  rcvd: 68
allowAny # 
```

numinit commented:
This looks good and should be fine to merge.

numinit commented on nixos/modules/services/networking/nebula.nix:
```diff
@@ -84,6 +84,28 @@ in
                 description = "Whether this node is a relay.";
               };
 
+              lighthouse.serveDns = lib.mkOption {
```
May be a good idea to move this to `lighthouse.dns.enable` so it's all under the same set of options.

siriobalmelli replied:
ok, respun

numinit commented on nixos/modules/services/networking/nebula.nix:
```diff
@@ -231,6 +253,19 @@ in
           );
+          capabilities =
+            let
+              nebulaPort = if settings.tun.disabled then settings.listen.port else 0;
```
This should  be `(!settings.tun.disabled)`

numinit reviewed (changes requested):
Missed the not on disabled, that will be a bug :-)

numinit commented:
@siriobalmelli I think this is good except for that one bug :-)

siriobalmelli commented:
Apologies, I was on leave. Respinnning and testing.

siriobalmelli committed:
nixos/nebula: add DNS options

Add the options:

- lighthouse.serve_dns
- lighthouse.dns.host
- lighthouse.dns.port

Improve systemd capabilities handling:

- do not give CAP_NET_ADMIN when tunnel interface is disabled
- give CAP_NET_BIND_SERVICE when DNS is enabled

Add self as maintainer: I'm using Nebula on NixOS in prod.

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

siriobalmelli committed:
nixos/nebula: add DNS tests

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

numinit reviewed (approved):
LGTM, tests run just fine.

numinit commented:
Howdy. If you re-push, we likely have to re-test. I don't currently have merge, but will try to get someone who does.

siriobalmelli commented:
> Howdy. If you re-push, we likely have to re-test. I don't currently have merge, but will try to get someone who does.

oh, derp.

I have a script which rebases all my open branches on top of `master`, and this of course is no exception.

I can try to exclude PR branches from the script, but then I'd effectively be maintaining two branches for each I/P PR ... since I dogfood these PRs and e.g. am actually running Nebula in production so I need my version of the changes to be rebased on top of `master` so I can update my systems.

I haven't had trouble with this in the past, usually things eventually get merged.

@numinit do you have any advice for me on how to best deal with this without causing annoyance to anyone else in the review process? Is this just a case of needing to get someone's attention be able to merge?

numinit commented:
Basically, yeah. GH isn't exactly obvious with what changed. I'd have merged it since it looked good, but my commit is pending.

siriobalmelli commented:
Ok, will hold the updates until this goes through 🤞

siriobalmelli commented:
Bump. Any chance this can get merged? 😄

numinit commented:
@rosscomputerguy another one for you

RossComputerGuy commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 353665`

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

RossComputerGuy reviewed (approved):
Had to run the tests myself and it succeeded.
![image0](https://github.com/user-attachments/assets/7dbcd236-370a-49fc-82ff-4f5529c129f9)
